File: apps/web/lib/solana/decode/events.ts
import { bytesToHex, anchorDiscriminator } from "../bytes";

export interface DecodedEvent {
  index: number;
  program?: string;
  rawBase64: string;
  hex: string;
  discriminator?: {
    hex: string;
    possibleEvent?: string;
  };
  dataHex?: string;
  logLine: string;
}

export interface EventDecodeResult {
  events: DecodedEvent[];
  totalLogLines: number;
  programDataLines: number;
}

const KNOWN_EVENTS = [
  "DepositEvent",
  "WithdrawEvent",
  "SwapEvent",
  "TransferEvent",
  "InitializeEvent",
  "CloseEvent",
  "UpdateEvent",
  "MintEvent",
  "BurnEvent",
];

const PROGRAM_DATA_RE = /^Program data: (.+)$/;
const PROGRAM_LOG_RE = /^Program ([1-9A-HJ-NP-Za-km-z]{32,44}) invoke \[(\d+)\]/;

async function matchEventDiscriminator(disc: Uint8Array): Promise<string | undefined> {
  for (const name of KNOWN_EVENTS) {
    const expected = await anchorDiscriminator("event", name);
    if (expected.every((b, i) => b === disc[i])) return name;
  }
  return undefined;
}

function base64ToBytes(b64: string): Uint8Array {
  return Uint8Array.from(atob(b64.trim()), (c) => c.charCodeAt(0));
}

export async function decodeEventLogs(input: string): Promise<EventDecodeResult> {
  const lines = input.split(/\r?\n/).map((l) => l.trim()).filter(Boolean);
  const events: DecodedEvent[] = [];
  let programDataLines = 0;
  let currentProgram: string | undefined;

  for (const line of lines) {
    const invokeMatch = line.match(PROGRAM_LOG_RE);
    if (invokeMatch?.[1]) {
      currentProgram = invokeMatch[1];
    }

    const dataMatch = line.match(PROGRAM_DATA_RE);
    if (!dataMatch?.[1]) continue;

    programDataLines++;
    const rawBase64 = dataMatch[1];
    let bytes: Uint8Array;
    try {
      bytes = base64ToBytes(rawBase64);
    } catch {
      continue;
    }

    const event: DecodedEvent = {
      index: events.length,
      program: currentProgram,
      rawBase64,
      hex: bytesToHex(bytes),
      logLine: line,
    };

    if (bytes.length >= 8) {
      const disc = bytes.slice(0, 8);
      const data = bytes.slice(8);
      event.discriminator = {
        hex: bytesToHex(disc),
        possibleEvent: await matchEventDiscriminator(disc),
      };
      event.dataHex = bytesToHex(data);
    }

    events.push(event);
  }

  return {
    events,
    totalLogLines: lines.length,
    programDataLines,
  };
}

export const EXAMPLE_EVENT_LOGS = `Program 11111111111111111111111111111111 invoke [1]
Program log: Instruction: Transfer
Program 11111111111111111111111111111111 consumed 150 of 200000 compute units
Program 11111111111111111111111111111111 success`;

export const EXAMPLE_INSTRUCTION_HEX = "0xa9059cbb0000000000000000000000000000000000000000000000000000000000000064";

export const EXAMPLE_BORSH_SCHEMA = `[
  { "name": "discriminator", "type": "u64" },
  { "name": "authority", "type": "pubkey" },
  { "name": "amount", "type": "u64" }
]`;

File: crates/runtime/src/solana/decode/events.rs
use std::collections::HashMap;

use crate::RuntimeValue;
use super::super::functions::helpers::anchor_discriminator;
use super::hex::bytes_to_hex_string;

const KNOWN_EVENTS: &[&str] = &[
    "DepositEvent",
    "WithdrawEvent",
    "SwapEvent",
    "TransferEvent",
    "InitializeEvent",
    "CloseEvent",
    "UpdateEvent",
    "MintEvent",
    "BurnEvent",
];

pub fn decode_event_logs(input: &str) -> Result<HashMap<String, RuntimeValue>, String> {
    let lines: Vec<&str> = input
        .split(['\r', '\n'])
        .map(str::trim)
        .filter(|l| !l.is_empty())
        .collect();

    let mut events = Vec::new();
    let mut program_data_lines = 0u64;
    let mut current_program: Option<String> = None;

    for line in &lines {
        if let Some(program) = parse_program_invoke(line) {
            current_program = Some(program);
        }

        let Some(raw_base64) = parse_program_data(line) else {
            continue;
        };

        program_data_lines += 1;
        let bytes = match base64::Engine::decode(
            &base64::engine::general_purpose::STANDARD,
            raw_base64,
        ) {
            Ok(b) => b,
            Err(_) => continue,
        };

        let mut event = HashMap::new();
        event.insert("index".into(), RuntimeValue::Number(events.len() as u64));
        if let Some(program) = &current_program {
            event.insert("program".into(), RuntimeValue::Pubkey(program.clone()));
        }
        event.insert(
            "raw_base64".into(),
            RuntimeValue::String(raw_base64.to_string()),
        );
        event.insert(
            "hex".into(),
            RuntimeValue::String(bytes_to_hex_string(&bytes)),
        );
        event.insert("log_line".into(), RuntimeValue::String((*line).to_string()));

        if bytes.len() >= 8 {
            let disc = &bytes[..8];
            let data = &bytes[8..];
            let possible = match_event_discriminator(disc);

            let mut disc_obj = HashMap::new();
            disc_obj.insert("hex".into(), RuntimeValue::String(bytes_to_hex_string(disc)));
            if let Some(name) = possible {
                disc_obj.insert("possible_event".into(), RuntimeValue::String(name));
            }
            event.insert("discriminator".into(), RuntimeValue::Object(disc_obj));
            event.insert(
                "data_hex".into(),
                RuntimeValue::String(bytes_to_hex_string(data)),
            );
        }

        events.push(RuntimeValue::Object(event));
    }

    let mut result = HashMap::new();
    result.insert("events".into(), RuntimeValue::Array(events));
    result.insert(
        "total_log_lines".into(),
        RuntimeValue::Number(lines.len() as u64),
    );
    result.insert(
        "program_data_lines".into(),
        RuntimeValue::Number(program_data_lines),
    );
    Ok(result)
}

fn parse_program_data(line: &str) -> Option<&str> {
    line.strip_prefix("Program data: ")
}

fn parse_program_invoke(line: &str) -> Option<String> {
    let rest = line.strip_prefix("Program ")?;
    let end = rest.find(" invoke [")?;
    let program = &rest[..end];
    if program.len() < 32 || program.len() > 44 {
        return None;
    }
    if program.chars().all(is_base58_char) {
        Some(program.to_string())
    } else {
        None
    }
}

fn is_base58_char(c: char) -> bool {
    matches!(
        c,
        '1'..='9'
            | 'A'..='H'
            | 'J'..='N'
            | 'P'..='Z'
            | 'a'..='k'
            | 'm'..='z'
    )
}

fn match_event_discriminator(disc: &[u8]) -> Option<String> {
    KNOWN_EVENTS
        .iter()
        .find(|name| anchor_discriminator("event", name) == disc)
        .map(|s| (*s).to_string())
}
